pkg/capabilities/capabilities.go

```diff
@@ -0,0 +1,263 @@
+	idMaxLength = 128
+)
+
+// NewCapabilityInfo returns a new CapabilityInfo.
+func NewCapabilityInfo(
+	id string,
+	capabilityType CapabilityType,
+	description string,
+	version string,
+) (CapabilityInfo, error) {
+	if len(id) > idMaxLength {
+		return CapabilityInfo{}, fmt.Errorf("invalid id: %s exceeds max length %d", id, idMaxLength)
+	}
+	if !idRegex.MatchString(id) {
+		return CapabilityInfo{}, fmt.Errorf("invalid id: %s. Allowed: %s", id, idRegex)
+	}
+
+	if ok := semver.IsValid(version); !ok {
+		return CapabilityInfo{}, fmt.Errorf("invalid version: %+v", version)
+	}
+
+	if err := capabilityType.IsValid(); err != nil {
+		return CapabilityInfo{}, err
+	}
+
+	return CapabilityInfo{
+		ID:             id,
+		CapabilityType: capabilityType,
+		Description:    description,
+		Version:        version,
+	}, nil
+}
+
+// MustNewCapabilityInfo returns a new CapabilityInfo,
+// panicking if we could not instantiate a CapabilityInfo.
+func MustNewCapabilityInfo(
+	id string,
+	capabilityType CapabilityType,
+	description string,
+	version string,
+) CapabilityInfo {
+	c, err := NewCapabilityInfo(id, capabilityType, description, version)
+	if err != nil {
+		panic(err)
+	}
+
+	return c
+}
+
+// TODO: this timeout was largely picked arbitrarily.
+// Consider what a realistic/desirable value should be.
+// See: https://smartcontract-it.atlassian.net/jira/software/c/projects/KS/boards/182
+var defaultExecuteTimeout = 10 * time.Second
+
+// ExecuteSync executes a capability synchronously.
+// We are not handling a case where a capability panics and crashes.
+// There is default timeout of 10 seconds. If a capability takes longer than
+// that then it should be executed asynchronously.
+func ExecuteSync(ctx context.Context, c CallbackExecutable, request CapabilityRequest) (*values.List, error) {
+	ctxWithT, cancel := context.WithTimeout(ctx, defaultExecuteTimeout)
+	defer cancel()
+
+	responseCh := make(chan CapabilityResponse)
+	setupCh := make(chan error)
+
+	go func(innerCtx context.Context, innerC CallbackExecutable, innerReq CapabilityRequest, innerCallback chan CapabilityResponse, errCh chan error) {
+		setupErr := innerC.Execute(innerCtx, innerCallback, innerReq)
+		setupCh <- setupErr
+	}(ctxWithT, c, request, responseCh, setupCh)
+
+	vs := make([]values.Value, 0)
+outerLoop:
+	for {
+		select {
+		case response, isOpen := <-responseCh:
+			if !isOpen {
+				break outerLoop
+			}
+			// An error means execution has been interrupted.
+			// We'll return the value discarding values received
+			// until now.
+			if response.Err != nil {
+				return nil, response.Err
+			}
+
+			vs = append(vs, response.Value)
+
+		// Timeout when a capability panics, crashes, and does not close the channel.
+		case <-ctxWithT.Done():
+			return nil, fmt.Errorf("context timed out. If you did not set a timeout, be aware that the default ExecuteSync timeout is %f seconds", defaultExecuteTimeout.Seconds())
+		}
+	}
+
+	setupErr := <-setupCh
```

Comment:
why isn't this in the select loop? as i read it, if there is a setup error, you waste 10s waiting for something that can't happen